Fire PushPredicateIntoTableScan rule with change in TableHandle

phd3 opened the pull request:
If there's a constraint in the filter node that the connector can "enforce", PushPredicateIntoTableScan modifies the plan to incorporate the new scan and residue filter.

In case of non-partition pushdown for hive, this is not true. The "compactEffectivePredicate" added to the table handle may or may not be satisfied by the HivePageSource (eg. ORC vs Avro). So we cannot get rid of those predicates from the FilterNode. But even though `HiveMetadata::applyFilter` returns a ConstraintApplicationResult, PushPredicateIntoTableScan throws it away, since `arePlansSame` doesn't consider change in table handles, if the enforced constraint hasn't changed. eg. Test added in `TestConnectorPushdownRulesWithHive` fails without this change. 

Predicates on top level columns still get pushed down to ORC/Parquet without this change, because `AddExchanges::visitFilter` adds the predicates, but I think the predicates should get pushed down in PushPredicateIntoTableScan.

I added connector table handle comparison in `arePlansSame`. (That failed some tests initially due to the fact that BucketFilter equality was object reference based, now they pass). But I'm not sure if comparing connector table handles is okay, or we need a better approach to tell that there was "some level" of predicate pushdown, even if it's not guaranteed to be enforced.

This change is also required for #1720 and #2672 to work together to pushdown predicates on nested columns.

martint commented on presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java:
```diff
@@ -131,7 +131,8 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
 
-        return Objects.equals(tableScan.getEnforcedConstraint(), rewrittenTableScan.getEnforcedConstraint());
+        return Objects.equals(tableScan.getEnforcedConstraint(), rewrittenTableScan.getEnforcedConstraint()) &&
+                Objects.equals(tableScan.getTable().getConnectorHandle(), rewrittenTableScan.getTable().getConnectorHandle());
```
This should compare the table handles directly. This layer of the code (i.e., the optimizer) shouldn't be concerned about ConnectorTableHandle.

phd3 commented:
I've verified equals and hashCode methods for all `ConnectorTableHandle` implementations, and made modifications where required to incorporate additional fields for comparison. 

Also added TableHandle equality method to address https://github.com/prestosql/presto/pull/3470#discussion_r411829525. Using TableHandle comparison requires `ConnectorTableLayoutHandle` to be compared too. we want to get rid of the usage altogether now, but I think it can be taken up in a separate PR.  

I've added the changes as separate commits right now for ease of review, will coalesce them into one if this feels noisy.
